README.md
<div align="center">

<img alt="gif-header" src="https://aemebe.weebly.com/uploads/4/7/1/4/47145535/tumblr-oujnx1jo9i1r9ndy3o1-500_orig.gif" width="350px"/>

<h2>✨ Dummy port scanner ✨</h2>

[![Semantic Release](https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg)](https://github.com/lpmatos/dummy-port-scanner)
[![Commitizen friendly](https://img.shields.io/badge/commitizen-friendly-brightgreen.svg)](https://github.com/lpmatos/dummy-port-scanner)
[![GitHub repo size](https://img.shields.io/github/repo-size/lpmatos/dummy-port-scanner)](https://github.com/lpmatos/dummy-port-scanner)

---

<img alt="gif-about" src="https://i.pinimg.com/originals/8b/e4/ef/8be4efc0a8e5bc4903aae00db82cb982.gif" width="450px" float="center"/>

<p>✨ Port Scanner is a simple CLI application designed to search servers or hosts for open ports ✨</p>

<p>
  <a href="#getting-started">Getting Started</a>&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;
  <a href="#concepts">Concepts</a>&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;
  <a href="#demo">Demo</a>&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;
  <a href="#versioning">Versioning</a>
</p>

</div>

---

## ➤ Getting Started <a name = "getting-started"></a>

If you want contribute on this project, first you need to make a **git clone**:

```bash
git clone --depth 1 https://github.com/lpmatos/dummy-port-scanner.git -b main
```

This will give you access to the code on your **local machine**.

## ➤ Important Note

**This tool is for educational and ethical practices only. The developers are not responsible if the tool is misused by an individual.**

## ➤ Installation <a name = "installation"></a>

```bash
python3 -m pip install -r requirements.txt
```

or

```bash
docker pull ghcr.io/lpmatos/dummy-port-scanner:main
```

## ➤ Usage <a name = "usage"></a>


```txt
$ python3 main.py --help

NAME
    main.py

SYNOPSIS
    main.py <flags>

FLAGS
    --target_host=TARGET_HOST
        Type: str
        Default: 'google.com'
    --target_ports=TARGET_PORTS
        Type: typing.List[int]
        Default: [443, 80]
    --mode=MODE
        Type: str
        Default: 'simple'
```

## ➤ Docker <a name = "docker"></a>

To run the docker container:

```bash
docker container run -it --rm ghcr.io/lpmatos/dummy-port-scanner:main
```

Click [here](https://github.com/lpmatos/dummy-port-scanner/pkgs/container/dummy-port-scanner/versions) to see available image tags.

## ➤ Features

- Pretty CLI.
- Docker container option.
- Scan target in a range of ports -  Loop mode.
- Scan target given a list of ports - Simple mode.
- Scan target given a list of ports using Nmap - Nmap mode.

## ➤ Concepts <a name = "concepts"></a>

This section aims to describe at a high level what the tools we use are and how we use them, without reproducing documentation that is better written (and more up to date) in the repositories and websites of these tools themselves. It's recommended to familiarize yourself with these tools as early as possible.

### Socket Programming

A socket uniquely identifies the endpoint of a communication link between two application ports. Socket programming is a way of connecting two nodes on a network to communicate with each other. One socket(node) listens on a particular port at an IP, while the other socket reaches out to the other to form a connection. The server forms the listener socket while the client reaches out to the server. They are the real backbones behind web browsing. In simpler terms, there is a server and a client.

A port represents an application process on a TCP/IP host, but the port number itself does not indicate the protocol being used: TCP, UDP, or IP. The application process might use the same port number for TCP or UDP protocols. To uniquely identify the destination of an IP packet arriving over the network, you have to extend the port principle with information about the protocol used and the IP address of the network interface; this information is called a socket. A socket has three parts: protocol, local-address, local-port.

The term association is used to specify completely the two processes that comprise a connection: protocol, local-address, local-port, foreign-address, foreign-port.

The terms socket and port are sometimes used as synonyms, but note that the terms port number and socket address are not like one another. A port number is one of the three parts of a socket address, and can be represented by a single number (for example, 1028) while a socket address can be represented by: tcp, myhostname, 1028.

### TCP/IP

Conceptually, the TCP/IP protocol stack consists of four layers, each layer consisting of one or more protocols. A protocol is a set of rules or standards that two entities must follow so as to allow each other to receive and interpret messages sent to them. The entities could, for example, be two application programs in an application protocol, or the entities might be two TCP protocol layers in two different IP hosts (the TCP protocol).

<p align="center">
  <img alt="gif-header" src="https://www.ibm.com/docs/en/SSLTBW_2.2.0/com.ibm.zos.v2r2.hala001/dwsl0044.gif" width="450px" float="center"/>
</p>

Programs are located at the process layer; here they can interface with the two transport layer protocols (TCP and UDP), or directly with the network layer protocols (ICMP and IP).

* TCP: Transmission Control Protocol is a transport protocol providing a reliable, full-duplex byte stream. Most TCP/IP applications use the TCP transport protocol.
* UDP: User Datagram Protocol is a connectionless protocol providing datagram services. UDP is less reliable because there is no guarantee that a UDP datagram ever reaches its intended destination, or that it reaches its destination only once and in the same condition as it was passed to the sending UDP layer by a UDP application.
* ICMP: Internet Control Message Protocol is used to handle error and control information at the IP layer. The ICMP is most often used by network control applications that are part of the TCP/IP software product itself, but ICMP can be used by authorized user processes as well. PING and TRACEROUTE are examples of network control applications that use the ICMP protocol.
* IP: Internet Protocol provides the packet delivery services for TCP, UDP, and ICMP. The IP layer protocol is unreliable (called a best-effort protocol). There is no guarantee that IP packets arrive, or that they arrive only once and are error-free. Such reliability is built into the TCP protocol, but not into the UDP protocol. If you need reliable transport between two UDP applications, you must ensure that reliability is built into the UDP applications.
* ARP: The IPv4 networking layer uses the Address Resolution Protocol (ARP) to map an IP address into a hardware address. In the IPv6 networking layer, this mapping is performed by the Neighbor Discovery (ND function). On local area networks (LANs), such an address would be called a media access control (MAC) address.
* RARP: Reverse Address Resolution Protocol is used to reverse the operation of the ARP protocol. It maps a hardware address into an IPv4 address. Note that both ARP packets and RARP packets are not forwarded in IP packets, but are themselves media level packets. ARP and RARP are not used on all network types, as some networks do not need these protocols.

## ➤ Demo <a name = "demo"></a>

<p align="center">
  <img alt="gif-header" src="./assets/demo.gif" width="750px" float="center"/>
</p>

## ➤ TL;DR

- Socket is an abstraction of an IP connection endpoint - so if you think of it as an API structure, you are not very far off.
- Internet layer i.e. IP Protocol. In practice you usually use explicitly sockets that bind to a certain transport layer parameters (datagram/UDP or stream/TCP)
- Sockets send data, in network byte order - whether it is text or binary, depends on the upper layer protocol.
- Theoretically, probably yes - but in practice all IP traffic is done using 'sockets'

## ➤ Author <a name = "author"></a>

👤 Hey!! If you like this project or if you find some bugs feel free to contact me in my channels:

>
> * Linktree: https://linktr.ee/lpmatos
>

## ➤ Versioning <a name = "versioning"></a>

To check the change history, please access the [**CHANGELOG.md**](CHANGELOG.md) file.

## ➤ Project status <a name = "project-status"></a>

This repository is a study project, therefore, it will not always be maintained 👻.

## ➤ Donations <a name = "donations"></a>

<p align="center">
  <a href="https://www.blockchain.com/pt/btc/address/bc1qn50elv826qs2qd6xhfh6n79649epqyaqmtwky5">
    <img alt="BTC Address" src="https://img.shields.io/badge/BTC%20Address-black?style=for-the-badge&logo=bitcoin&logoColor=white">
  </a>

  <a href="https://live.blockcypher.com/ltc/address/ltc1qwzrxmlmzzx68k2dnrcrplc4thadm75khzrznjw/">
    <img alt="Litecoin Address" src="https://img.shields.io/badge/Litecoin%20Address-black?style=for-the-badge&logo=litecoin&logoColor=white">
  </a>
</p>

## ➤ Show your support <a name = "show-your-support"></a>

<div align="center">

Give me a ⭐️ if this project helped you!

<img alt="gif-header" src="https://www.icegif.com/wp-content/uploads/baby-yoda-bye-bye-icegif.gif" width="350px" float="center"/>

Made with 💜 by [me](https://github.com/lpmatos) 👋 inspired on [readme-md-generator](https://github.com/kefranabg/readme-md-generator)

</div>


CHANGELOG.md
# Semantic Versioning Changelog

### [1.3.2](https://github.com/lpmatos/dummy-port-scanner/compare/1.3.1...1.3.2) (2022-04-11)


### :repeat: CI

* add --extends in semantic release ([de9c4e6](https://github.com/lpmatos/dummy-port-scanner/commit/de9c4e63bc279a98f4d257de1ad6a7db2244bc15))


### :bug: Fixes

* cp file to release ([a639ac4](https://github.com/lpmatos/dummy-port-scanner/commit/a639ac44aa50780faf3a129dffeed2d7a5b389b4))
* merge conflicts ([7c9b077](https://github.com/lpmatos/dummy-port-scanner/commit/7c9b07729f2ba3802879ede24bf090afd9fd6652))

### [1.3.1](https://github.com/lpmatos/dummy-port-scanner/compare/1.3.0...1.3.1) (2022-04-11)


### :bug: Fixes

* dockerfile security package version ([4cd4579](https://github.com/lpmatos/dummy-port-scanner/commit/4cd457920a08ffdeea7869b8ea89c0625760223f))

## [1.3.0](https://github.com/lpmatos/dummy-port-scanner/compare/1.2.0...1.3.0) (2022-04-11)


### :memo: Docs

* add donations ([60d1925](https://github.com/lpmatos/dummy-port-scanner/commit/60d192572707981c353e1f6aba6058c3dd2edaac))


### :sparkles: News

* add description args ([5d25777](https://github.com/lpmatos/dummy-port-scanner/commit/5d257772a5925562cb041647809e8442b4ca9f48))


### :bug: Fixes

* dockerfile and precommit setup ([dcfd5d0](https://github.com/lpmatos/dummy-port-scanner/commit/dcfd5d09e318a7833de86d2f76b805a2778a34f6))
* dockerfile hadolint ([a35d958](https://github.com/lpmatos/dummy-port-scanner/commit/a35d9588931762f383d6cda04138776980e0b6f7))
* files organization ([e94e58c](https://github.com/lpmatos/dummy-port-scanner/commit/e94e58c9e64581001a72a4909e1e4eb6d7a52042))
* remove entrypoint setup docker-compose ([a7ebb32](https://github.com/lpmatos/dummy-port-scanner/commit/a7ebb32d85a23ea621566f56eeb87babd00e5aa6))
* rename ci files ([9cc131b](https://github.com/lpmatos/dummy-port-scanner/commit/9cc131bdfd8c7071215093ae00d26250a13f6961))

## [1.2.0](https://github.com/lpmatos/dummy-port-scanner/compare/1.1.0...1.2.0) (2022-01-06)


### :sparkles: News

* add durantion scan print ([80d6e3e](https://github.com/lpmatos/dummy-port-scanner/commit/80d6e3e5b1b313abcf5b1c009678a0df8ea2c3aa))
* add loop mode and default host google.com ([a48d434](https://github.com/lpmatos/dummy-port-scanner/commit/a48d434ebb991afe80319d3d7741f219f0046f20))
* add mode class ([6fe2dec](https://github.com/lpmatos/dummy-port-scanner/commit/6fe2dec390f700e43226cad3e3e3a26a4dc4bf7a))


### :bug: Fixes

* default dockerfile cmd ([5115f36](https://github.com/lpmatos/dummy-port-scanner/commit/5115f366916d5a492e7782c01eaa4712e25baada))


### :memo: Docs

* add comments about modes ([8926ac8](https://github.com/lpmatos/dummy-port-scanner/commit/8926ac8bd287e19ae4beec8ee85e3d88f48d1d78))
* add features and important note ([addcae6](https://github.com/lpmatos/dummy-port-scanner/commit/addcae629d33dfac46e8daa7b50dc3f69285f139))
* fix usage code ([0922a29](https://github.com/lpmatos/dummy-port-scanner/commit/0922a29afb8c2954f7a62a6f90fb4d5f9328b6c4))

## [1.1.0](https://github.com/lpmatos/dummy-port-scanner/compare/1.0.0...1.1.0) (2022-01-06)


### :repeat: CI

* add github workspace path ([55c43ec](https://github.com/lpmatos/dummy-port-scanner/commit/55c43ecd893adbe33e152002483d05d3a0364d30))
* add reference to trojan config file ([018bd31](https://github.com/lpmatos/dummy-port-scanner/commit/018bd31118824a1372efbf4442f3a46e7dab75cf))
* add trojan source detector ([05dfb55](https://github.com/lpmatos/dummy-port-scanner/commit/05dfb55d27e2a6ce5e78fbacc500ddaaa65c051d))
* fix hadolint with dockerfile ([75ec08a](https://github.com/lpmatos/dummy-port-scanner/commit/75ec08a794714db41a2a18672ef48bcc657321b1))
* fix reference config trojan ([cf3e9e2](https://github.com/lpmatos/dummy-port-scanner/commit/cf3e9e2a5457c2ed611e09abd515b58b716b1dcd))


### :bug: Fixes

* centralize config files and change call in templates ([8d4dcfa](https://github.com/lpmatos/dummy-port-scanner/commit/8d4dcfa6ff51b7b273636e3faedf7534f5ec65f3))
* remove semantic release exec config ([641527b](https://github.com/lpmatos/dummy-port-scanner/commit/641527b1d936f131e5746fa8328b7ea8859a07cf))
* remove source trojan def ([d20fa4b](https://github.com/lpmatos/dummy-port-scanner/commit/d20fa4bf9bea8159e1c570ade7d35fb401b05ef5))


### :sparkles: News

* add directory and exclude assets in trojan config ([9ef349a](https://github.com/lpmatos/dummy-port-scanner/commit/9ef349a3489cfc3f2642611259b5295179515c90))
* add docker extension gitpod ([620c84c](https://github.com/lpmatos/dummy-port-scanner/commit/620c84c912658435771e2cce362820dc3b13a280))
* add docker-compose ([a23784f](https://github.com/lpmatos/dummy-port-scanner/commit/a23784faaefd39eda0db88638999ca103719e94f))
* add gitpod dockerfile ([155e9e4](https://github.com/lpmatos/dummy-port-scanner/commit/155e9e4e4d1dbc50679da086c4df26493f80b7d9))
* add gitpod extensions ([1ef145f](https://github.com/lpmatos/dummy-port-scanner/commit/1ef145f8ea86c7d6d3727bae2f3f20c8b98af28f))
* config trojan file ([65826f2](https://github.com/lpmatos/dummy-port-scanner/commit/65826f2574a60e6af449546f1f7c5203f42654c0))


### :memo: Docs

* change description project ([2dac4af](https://github.com/lpmatos/dummy-port-scanner/commit/2dac4af6e3823e97d8792b538b43621d31082770))
* change sort topics menu ([1bb5845](https://github.com/lpmatos/dummy-port-scanner/commit/1bb58458927d953d1dc52f09c54dda0d8aba49af))
* fix getting started url ([7c77956](https://github.com/lpmatos/dummy-port-scanner/commit/7c77956d5f5eaa1aedaea8b1277179984d9f6299))

## 1.0.0 (2022-01-06)


### :sparkles: News

* initial commit - project organization ([42ded66](https://github.com/lpmatos/dummy-port-scanner/commit/42ded667cd16327837701ec05e5d5f3be233df92))


### :bug: Fixes

* remove changelog file ([89a69c5](https://github.com/lpmatos/dummy-port-scanner/commit/89a69c55dc64dce9a1bb1ba18c76df4dfd6345e7))


src/main.py
# -*- coding: utf-8 -*-

import socket, fire, pyfiglet, sys
from loguru import logger
from nmap import PortScanner
from typing import List
from datetime import datetime


class BColors:
    OKCYAN = "\033[96m"
    WARNING = "\033[93m"

class Mode:
    SIMPLE = "simple"
    LOOP = "loop"
    NMAP = "nmap"

def connect_scan(target_host: str, target_port: int) -> None:
    # AF_INET refers to the ipv4 family-address
    # SOCK_STREAM means that the connection is oriented by TCP protocol
    with socket.socket(family=socket.AF_INET, type=socket.SOCK_STREAM) as conn:
        try:
            conn.connect((target_host, target_port))
            logger.info(f"👻 Port {target_port} Opened")
        except:
            logger.warning(f"🥶 Port {target_port} Closed")
        finally:
            logger.debug("💥 Closing connection")
            conn.close()


def port_scan_simple(target_host: str, target_ports: List[int] = [443, 80]) -> None:
    logger.debug("✨ Enable port scan simple mode")
    for port in target_ports:
        logger.debug(f"🔎 Scanning port: {port}")
        connect_scan(target_host, int(port))


def nmap_scan(target_host: str, target_ports: List[int] = [443, 80]) -> None:
    logger.debug("✨ Enable port scan nmap mode")
    scan = PortScanner()
    for port in target_ports:
        print(f"Scanning port: {port}")
        portscan = scan.scan(target_host, str(port))
        print("Port", port, "is ", portscan["scan"][target_host]["tcp"][port]["state"])

def port_scan_loop(target_host: str) -> None:
    logger.debug("✨ Enable port scan loop mode")
    for port in range(1,52000):
        logger.debug(f"🔎 Scanning port: {port}")
        connect_scan(target_host, int(port))

def port_scan(
    target_host: str = "google.com",
    target_ports: List[int] = [443, 80],
    mode: str = "simple"
) -> None:

    print(BColors.OKCYAN + f"✨ Target host: {target_host}")
    print(BColors.OKCYAN + f"✨ Target ports: {target_ports}")
    print(BColors.OKCYAN + f"✨ Target mode: {mode}", end="\n\n")

    try:
        target_ip = socket.gethostbyname(target_host)
    except:
        print(f"🤬 Cannot resolve {target_host}")
        return

    try:
        target_name = socket.gethostbyaddr(target_ip)
        logger.debug(f"👾 Scan result of {target_name[0]}")
    except:
        logger.debug(f"👾 Scan result of {target_ip}")

    socket.setdefaulttimeout(1)

    start = datetime.now()

    if mode == Mode.SIMPLE:
        port_scan_simple(target_host, target_ports)

    if mode == Mode.LOOP:
        port_scan_loop(target_host)

    if mode == Mode.NMAP:
        nmap_scan(target_host, target_ports)

    end = datetime.now()
    print(f"\nScanning completed in: {end - start}\n")


def main() -> None:
    ascii_banner = pyfiglet.figlet_format("Port Scanner")
    print(BColors.OKCYAN + ascii_banner)
    print(BColors.WARNING + "                   by @CI Monk")
    print(BColors.WARNING + "		      @lpmatos \n")
    fire.Fire(port_scan)


if __name__ == "__main__":
    main()
